File: packages/telemetry/typescript/src/sdk/memory.ts
import { type Attributes, context as otelContext, SpanKind, SpanStatusCode, trace } from "@opentelemetry/api"
import { GEN_AI_MEMORY_ATTRIBUTES, type MEMORY_OPERATIONS } from "../constants/attributes.ts"
import type { Latitude } from "./init.ts"
import { latitudeAttributesFromContext } from "./tracer.ts"
import type { ContextOptions } from "./types.ts"

type MaybePromise<T> = T | Promise<T>

export type MemoryOperation = (typeof MEMORY_OPERATIONS)[number]

/** A memory record as defined by the OTEL `gen_ai.memory.records` schema: `content` is required. */
export type MemoryRecordInput = {
  readonly content: unknown
  readonly id?: string
  readonly score?: number
  readonly metadata?: Record<string, unknown>
}

export type MemoryRedactInfo = {
  readonly operation: MemoryOperation
  readonly storeId: string
}

export type MemoryTelemetryOptions = {
  readonly latitude: Latitude
  /** Instrumentation scope suffix; defaults to `"memory"` → `so.latitude.instrumentation.memory`. */
  readonly scope?: string
  /** Default `gen_ai.memory.store.id` for every operation; per-call `storeId` overrides it. */
  readonly storeId?: string
  readonly context?: ContextOptions | (() => ContextOptions | undefined)
  /** Send record bodies (`gen_ai.memory.records`). Off by default — opt-in per OTEL and to avoid shipping PII. */
  readonly captureContent?: boolean
  readonly redact?: (records: readonly MemoryRecordInput[], info: MemoryRedactInfo) => readonly MemoryRecordInput[]
}

type MemoryOpBaseOptions = {
  readonly storeId?: string
  readonly count?: number
  readonly captureContent?: boolean
}

type MemoryWriteOptions = MemoryOpBaseOptions & {
  readonly recordId?: string
  readonly records?: readonly MemoryRecordInput[]
}

/** Omit `recordId` to signal a whole-store wipe. */
type MemoryDeleteOptions = MemoryOpBaseOptions & {
  readonly recordId?: string
}

type MemorySearchOptions = MemoryOpBaseOptions & {
  readonly query?: string
  readonly records?: readonly MemoryRecordInput[]
}

type MemoryStoreOptions = {
  readonly storeId?: string
}

type WithExecute<O, T> = O & { readonly execute: () => MaybePromise<T> }

type SearchWithExecute<T> = MemorySearchOptions & {
  readonly execute: () => MaybePromise<T>
  /** Map the search result to the records it returned so they attribute to this span. */
  readonly recordsFromResult?: (result: T) => readonly MemoryRecordInput[]
}

export interface MemoryTelemetry {
  create<T>(options: WithExecute<MemoryWriteOptions, T>): Promise<T>
  create(options?: MemoryWriteOptions): Promise<void>
  update<T>(options: WithExecute<MemoryWriteOptions, T>): Promise<T>
  update(options?: MemoryWriteOptions): Promise<void>
  upsert<T>(options: WithExecute<MemoryWriteOptions, T>): Promise<T>
  upsert(options?: MemoryWriteOptions): Promise<void>
  delete<T>(options: WithExecute<MemoryDeleteOptions, T>): Promise<T>
  delete(options?: MemoryDeleteOptions): Promise<void>
  search<T>(options: SearchWithExecute<T>): Promise<T>
  search(options?: MemorySearchOptions): Promise<void>
  createStore<T>(options: WithExecute<MemoryStoreOptions, T>): Promise<T>
  createStore(options?: MemoryStoreOptions): Promise<void>
  deleteStore<T>(options: WithExecute<MemoryStoreOptions, T>): Promise<T>
  deleteStore(options?: MemoryStoreOptions): Promise<void>
}

function stringifyAttribute(value: unknown): string {
  try {
    return JSON.stringify(value)
  } catch {
    return String(value)
  }
}

type RunParams<T> = {
  readonly operation: MemoryOperation
  readonly storeId?: string
  readonly recordId?: string
  readonly count?: number
  readonly query?: string
  readonly records?: readonly MemoryRecordInput[]
  readonly captureContent?: boolean
  readonly execute?: () => MaybePromise<T>
  readonly recordsFromResult?: (result: T) => readonly MemoryRecordInput[]
}

export function createMemoryTelemetry(options: MemoryTelemetryOptions): MemoryTelemetry {
  const scope = options.scope ?? "memory"
  const defaultStoreId = options.storeId
  const defaultCaptureContent = options.captureContent ?? false
  const redact = options.redact ?? ((records: readonly MemoryRecordInput[]) => records)

  const resolveContext = (): ContextOptions | undefined =>
    typeof options.context === "function" ? options.context() : options.context

  const buildAttributes = (params: {
    operation: MemoryOperation
    storeId: string
    recordId?: string | undefined
    count?: number | undefined
    query?: string | undefined
    records?: readonly MemoryRecordInput[] | undefined
    captureContent: boolean
    context: ContextOptions | undefined
  }): Attributes => {
    const attributes: Attributes = {
      ...latitudeAttributesFromContext(params.context ?? {}),
      [GEN_AI_MEMORY_ATTRIBUTES.operationName]: params.operation,
      [GEN_AI_MEMORY_ATTRIBUTES.storeId]: params.storeId,
    }
    if (params.recordId) attributes[GEN_AI_MEMORY_ATTRIBUTES.recordId] = params.recordId
    const count = params.count ?? params.records?.length
    if (count !== undefined) attributes[GEN_AI_MEMORY_ATTRIBUTES.recordCount] = count
    if (params.query) attributes[GEN_AI_MEMORY_ATTRIBUTES.queryText] = params.query
    if (params.captureContent && params.records && params.records.length > 0) {
      const redacted = redact(params.records, { operation: params.operation, storeId: params.storeId })
      attributes[GEN_AI_MEMORY_ATTRIBUTES.records] = stringifyAttribute(redacted)
    }
    return attributes
  }

  const run = <T>(params: RunParams<T>): Promise<T | undefined> => {
    const context = resolveContext()
    const storeId = params.storeId ?? defaultStoreId ?? ""
    const captureContent = params.captureContent ?? defaultCaptureContent
    const attributes = buildAttributes({
      operation: params.operation,
      storeId,
      recordId: params.recordId,
      count: params.count,
      query: params.query,
      records: params.records,
      captureContent,
      context,
    })
    const parentContext = otelContext.active()
    const span = options.latitude
      .getTracer(scope)
      .startSpan(params.operation, { kind: SpanKind.CLIENT, attributes }, parentContext)

    const execute = params.execute
    if (!execute) {
      span.end()
      return Promise.resolve(undefined)
    }

    return otelContext.with(trace.setSpan(parentContext, span), async () => {
      try {
        const result = await execute()
        if (params.recordsFromResult) {
          const derived = params.recordsFromResult(result)
          if (derived.length > 0) {
            span.setAttributes(
              buildAttributes({
                operation: params.operation,
                storeId,
                count: params.count ?? derived.length,
                query: params.query,
                records: derived,
                captureContent,
                context,
              }),
            )
          }
        }
        return result
      } catch (error) {
        span.recordException(error as Error)
        span.setStatus({ code: SpanStatusCode.ERROR, message: error instanceof Error ? error.message : String(error) })
        throw error
      } finally {
        span.end()
      }
    })
  }

  function create<T>(options: WithExecute<MemoryWriteOptions, T>): Promise<T>
  function create(options?: MemoryWriteOptions): Promise<void>
  function create<T>(opts: MemoryWriteOptions & { execute?: () => MaybePromise<T> } = {}): Promise<unknown> {
    return run<T>({ operation: "create_memory", ...opts })
  }

  function update<T>(options: WithExecute<MemoryWriteOptions, T>): Promise<T>
  function update(options?: MemoryWriteOptions): Promise<void>
  function update<T>(opts: MemoryWriteOptions & { execute?: () => MaybePromise<T> } = {}): Promise<unknown> {
    return run<T>({ operation: "update_memory", ...opts })
  }

  function upsert<T>(options: WithExecute<MemoryWriteOptions, T>): Promise<T>
  function upsert(options?: MemoryWriteOptions): Promise<void>
  function upsert<T>(opts: MemoryWriteOptions & { execute?: () => MaybePromise<T> } = {}): Promise<unknown> {
    return run<T>({ operation: "upsert_memory", ...opts })
  }

  function deleteOp<T>(options: WithExecute<MemoryDeleteOptions, T>): Promise<T>
  function deleteOp(options?: MemoryDeleteOptions): Promise<void>
  function deleteOp<T>(opts: MemoryDeleteOptions & { execute?: () => MaybePromise<T> } = {}): Promise<unknown> {
    return run<T>({ operation: "delete_memory", ...opts })
  }

  function search<T>(options: SearchWithExecute<T>): Promise<T>
  function search(options?: MemorySearchOptions): Promise<void>
  function search<T>(opts: SearchWithExecute<T> | MemorySearchOptions = {}): Promise<unknown> {
    return run<T>({ operation: "search_memory", ...opts } as RunParams<T>)
  }

  function createStore<T>(options: WithExecute<MemoryStoreOptions, T>): Promise<T>
  function createStore(options?: MemoryStoreOptions): Promise<void>
  function createStore<T>(opts: MemoryStoreOptions & { execute?: () => MaybePromise<T> } = {}): Promise<unknown> {
    return run<T>({ operation: "create_memory_store", ...opts })
  }

  function deleteStore<T>(options: WithExecute<MemoryStoreOptions, T>): Promise<T>
  function deleteStore(options?: MemoryStoreOptions): Promise<void>
  function deleteStore<T>(opts: MemoryStoreOptions & { execute?: () => MaybePromise<T> } = {}): Promise<unknown> {
    return run<T>({ operation: "delete_memory_store", ...opts })
  }

  return { create, update, upsert, delete: deleteOp, search, createStore, deleteStore }
}

File: packages/telemetry/python/src/latitude_telemetry/sdk/memory.py
"""
First-party helper for emitting OpenTelemetry GenAI memory-operation spans.

Mirrors the TypeScript `createMemoryTelemetry`: each operation optionally wraps an
`execute` callable (capturing latency, errors, and status) or emits a completed span.
Record content (`gen_ai.memory.records`) is opt-in via `capture_content`.
"""

import inspect
import json
from collections.abc import Callable, Sequence
from typing import TYPE_CHECKING, Any, NotRequired, Required, TypedDict

from opentelemetry import context as otel_context
from opentelemetry import trace
from opentelemetry.trace import SpanKind, Status, StatusCode
from opentelemetry.util.types import AttributeValue

from latitude_telemetry.constants import MEMORY_ATTRIBUTES
from latitude_telemetry.sdk.tracer import latitude_attributes_from_context
from latitude_telemetry.sdk.types import ContextOptions

if TYPE_CHECKING:
    from latitude_telemetry.sdk.init import Latitude


class MemoryRecord(TypedDict, total=False):
    """A record per the OTEL `gen_ai.memory.records` schema; `content` is required."""

    content: Required[object]
    id: NotRequired[str]
    score: NotRequired[float]
    metadata: NotRequired[dict[str, object]]


class MemoryRedactInfo(TypedDict):
    operation: str
    store_id: str


RedactFn = Callable[[list[MemoryRecord], MemoryRedactInfo], list[MemoryRecord]]
ContextInput = ContextOptions | Callable[[], ContextOptions | None] | None


def _identity_redact(records: list[MemoryRecord], info: MemoryRedactInfo) -> list[MemoryRecord]:
    return records


def _stringify(value: object) -> str:
    try:
        return json.dumps(value)
    except (TypeError, ValueError):
        return str(value)


class MemoryTelemetry:
    def __init__(
        self,
        latitude: "Latitude",
        *,
        scope: str = "memory",
        store_id: str | None = None,
        context: ContextInput = None,
        capture_content: bool = False,
        redact: RedactFn | None = None,
    ):
        self._latitude = latitude
        self._scope = scope
        self._default_store_id = store_id
        self._context = context
        self._default_capture_content = capture_content
        self._redact: RedactFn = redact or _identity_redact

    def _resolve_context(self) -> ContextOptions | None:
        return self._context() if callable(self._context) else self._context

    def _build_attributes(
        self,
        operation: str,
        store_id: str,
        record_id: str | None,
        count: int | None,
        query: str | None,
        records: Sequence[MemoryRecord] | None,
        capture_content: bool,
        context: ContextOptions | None,
    ) -> dict[str, AttributeValue]:
        attributes: dict[str, AttributeValue] = dict(latitude_attributes_from_context(context or {}))
        attributes[MEMORY_ATTRIBUTES.operation_name] = operation
        attributes[MEMORY_ATTRIBUTES.store_id] = store_id
        if record_id:
            attributes[MEMORY_ATTRIBUTES.record_id] = record_id
        resolved_count = count if count is not None else (len(records) if records else None)
        if resolved_count is not None:
            attributes[MEMORY_ATTRIBUTES.record_count] = resolved_count
        if query:
            attributes[MEMORY_ATTRIBUTES.query_text] = query
        if capture_content and records:
            redacted = self._redact(list(records), {"operation": operation, "store_id": store_id})
            attributes[MEMORY_ATTRIBUTES.records] = _stringify(redacted)
        return attributes

    def _run(
        self,
        operation: str,
        *,
        store_id: str | None = None,
        record_id: str | None = None,
        count: int | None = None,
        query: str | None = None,
        records: Sequence[MemoryRecord] | None = None,
        capture_content: bool | None = None,
        execute: Callable[[], Any] | None = None,
        records_from_result: Callable[[Any], Sequence[MemoryRecord]] | None = None,
    ) -> Any:
        context = self._resolve_context()
        sid = store_id if store_id is not None else (self._default_store_id or "")
        cc = capture_content if capture_content is not None else self._default_capture_content
        attributes = self._build_attributes(operation, sid, record_id, count, query, records, cc, context)
        span = self._latitude.get_tracer(self._scope).start_span(operation, kind=SpanKind.CLIENT, attributes=attributes)

        def _record_search(result: Any) -> None:
            if records_from_result is None:
                return
            derived = records_from_result(result)
            if derived:
                span.set_attributes(
                    self._build_attributes(
                        operation, sid, None, count if count is not None else len(derived), query, derived, cc, context
                    )
                )

        if execute is None:
            span.end()
            return None

        if inspect.iscoroutinefunction(execute):

            async def _async_run() -> Any:
                token = otel_context.attach(trace.set_span_in_context(span))
                try:
                    result = await execute()
                    _record_search(result)
                    return result
                except Exception as error:
                    span.record_exception(error)
                    span.set_status(Status(StatusCode.ERROR, str(error)))
                    raise
                finally:
                    span.end()
                    otel_context.detach(token)

            return _async_run()

        token = otel_context.attach(trace.set_span_in_context(span))
        try:
            result = execute()
        except Exception as error:
            span.record_exception(error)
            span.set_status(Status(StatusCode.ERROR, str(error)))
            span.end()
            otel_context.detach(token)
            raise

        if inspect.isawaitable(result):
            otel_context.detach(token)

            async def _await_result() -> Any:
                inner_token = otel_context.attach(trace.set_span_in_context(span))
                try:
                    awaited = await result
                    _record_search(awaited)
                    return awaited
                except Exception as error:
                    span.record_exception(error)
                    span.set_status(Status(StatusCode.ERROR, str(error)))
                    raise
                finally:
                    span.end()
                    otel_context.detach(inner_token)

            return _await_result()

        _record_search(result)
        span.end()
        otel_context.detach(token)
        return result

    def create(
        self,
        *,
        store_id: str | None = None,
        record_id: str | None = None,
        records: Sequence[MemoryRecord] | None = None,
        count: int | None = None,
        capture_content: bool | None = None,
        execute: Callable[[], Any] | None = None,
    ) -> Any:
        return self._run(
            "create_memory",
            store_id=store_id,
            record_id=record_id,
            records=records,
            count=count,
            capture_content=capture_content,
            execute=execute,
        )

    def update(
        self,
        *,
        store_id: str | None = None,
        record_id: str | None = None,
        records: Sequence[MemoryRecord] | None = None,
        count: int | None = None,
        capture_content: bool | None = None,
        execute: Callable[[], Any] | None = None,
    ) -> Any:
        return self._run(
            "update_memory",
            store_id=store_id,
            record_id=record_id,
            records=records,
            count=count,
            capture_content=capture_content,
            execute=execute,
        )

    def upsert(
        self,
        *,
        store_id: str | None = None,
        record_id: str | None = None,
        records: Sequence[MemoryRecord] | None = None,
        count: int | None = None,
        capture_content: bool | None = None,
        execute: Callable[[], Any] | None = None,
    ) -> Any:
        return self._run(
            "upsert_memory",
            store_id=store_id,
            record_id=record_id,
            records=records,
            count=count,
            capture_content=capture_content,
            execute=execute,
        )

    def delete(
        self,
        *,
        store_id: str | None = None,
        record_id: str | None = None,
        count: int | None = None,
        capture_content: bool | None = None,
        execute: Callable[[], Any] | None = None,
    ) -> Any:
        """Omit `record_id` to signal a whole-store wipe."""
        return self._run(
            "delete_memory",
            store_id=store_id,
            record_id=record_id,
            count=count,
            capture_content=capture_content,
            execute=execute,
        )

    def search(
        self,
        *,
        store_id: str | None = None,
        query: str | None = None,
        records: Sequence[MemoryRecord] | None = None,
        count: int | None = None,
        capture_content: bool | None = None,
        execute: Callable[[], Any] | None = None,
        records_from_result: Callable[[Any], Sequence[MemoryRecord]] | None = None,
    ) -> Any:
        return self._run(
            "search_memory",
            store_id=store_id,
            query=query,
            records=records,
            count=count,
            capture_content=capture_content,
            execute=execute,
            records_from_result=records_from_result,
        )

    def create_store(self, *, store_id: str | None = None, execute: Callable[[], Any] | None = None) -> Any:
        return self._run("create_memory_store", store_id=store_id, execute=execute)

    def delete_store(self, *, store_id: str | None = None, execute: Callable[[], Any] | None = None) -> Any:
        return self._run("delete_memory_store", store_id=store_id, execute=execute)


def create_memory_telemetry(
    latitude: "Latitude",
    *,
    scope: str = "memory",
    store_id: str | None = None,
    context: ContextInput = None,
    capture_content: bool = False,
    redact: RedactFn | None = None,
) -> MemoryTelemetry:
    return MemoryTelemetry(
        latitude,
        scope=scope,
        store_id=store_id,
        context=context,
        capture_content=capture_content,
        redact=redact,
    )
